### PR TITLE
JSON "null" request body rejected as an empty request body

### LINKED ISSUE #40415: JSON "null" request body rejected as an empty request body
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

I have an action method which accepts a posted `null` value for the body:

```
		[HttpPost("acceptsNull")]
		public string Upload(WeatherForecast? forecast = null)
		{
			return JsonConvert.SerializeObject(forecast);
		}
```

However, when I post a JSON null to this endpoint I get a 400 response complaining that the body is empty.

### Expected Behavior

Since the body is not empty and JSON-deserializes to the model parameter type just fine, I expect it to invoke the action with a null model.

If it can't do that, I at least expect an error message that indicates nulls are not permitted rather than one claiming the body is empty (although if the body were actually empty instead of valid JSON I'd probably want this request to be rejected).

### Steps To Reproduce

Call the above endpoint with:

```
using var client = new HttpClient();
using var response = client.PostAsJsonAsync(".../acceptsNull", default(WeatherForecast));
Console.WriteLine(response.StatusCode); // 400
Console.WriteLine(await response.Content.ReadAsStringAsync()); //  ...A non-empty request body is required....
```

### Exceptions (if any)

_No response_

### .NET Version

6.0.100

### Anything else?

The workaround I found is to set `[FromBody(EmptyBodyBehavior = EmptyBodyBehavior.Allow)]`, but I don't really want to have to do this because I'd like to reject empty bodies and accept non-empty bodies which are JSON null.

### PR DESCRIPTION
**Pull request by senbar**
JSON "null" response body rejected as an empty response body

I have added check if request body has 0 bytes; If yes then body was actually empty and proper error message is returned; If not then json contained null and different message is returned. 

Should I create new error message for this null case in resources? I have changed it to ValueMustNotBeNullAccessor but it doesn't seem appropriate since it was for a bit different cases elsewhere.

Fixes #40415

### CODE REVIEW

**Comment by senbar** [comment]
I think this should do it. If anyone has better idea how to check whether request was empty or with null value please let me know. Also the error message is up for debate I dont think the one I used is clear at all

File: src/Mvc/Mvc.Core/src/ModelBinding/Binders/BodyModelBinder.cs [hunk 1]
```diff
@@ -183,11 +183,24 @@ public async Task BindModelAsync(ModelBindingContext bindingContext)
...
                 // returning InputFormatterResult.Success(defaultForModelType), because input formatters
                 // are responsible for choosing a default value for the model type.
-                var message = bindingContext
-                    .ModelMetadata
-                    .ModelBindingMessageProvider
-                    .MissingRequestBodyRequiredValueAccessor();
-                bindingContext.ModelState.AddModelError(modelBindingKey, message);
+
+                // If request content length is 0 then request had empty body, if not null was provided in json
+                if (httpContext.Request.ContentLength == 0)
```
Comment (brunolins16) [inline src/Mvc/Mvc.Core/src/ModelBinding/Binders/BodyModelBinder.cs:188]: I don't feel that this is the right place to do this. Maybe an option is change the json-based inpur formatters to AddModelError and return Failure instead

https://github.com/dotnet/aspnetcore/blob/1c5da276ce957e9db063d7dc7440ea6cc496300c/src/Mvc/Mvc.NewtonsoftJson/src/NewtonsoftJsonInputFormatter.cs#L206

https://github.com/dotnet/aspnetcore/blob/1c5da276ce957e9db063d7dc7440ea6cc496300c/src/Mvc/Mvc.Core/src/Formatters/SystemTextJsonInputFormatter.cs#L111

Please check if we do have any side effects with this change.

Comment (brunolins16) [inline src/Mvc/Mvc.Core/src/ModelBinding/Binders/BodyModelBinder.cs:188]: I don't feel that this is the right place to do this. Maybe an option is change the json-based inpur formatters to AddModelError and return Failure instead

https://github.com/dotnet/aspnetcore/blob/1c5da276ce957e9db063d7dc7440ea6cc496300c/src/Mvc/Mvc.NewtonsoftJson/src/NewtonsoftJsonInputFormatter.cs#L206

https://github.com/dotnet/aspnetcore/blob/1c5da276ce957e9db063d7dc7440ea6cc496300c/src/Mvc/Mvc.Core/src/Formatters/SystemTextJsonInputFormatter.cs#L111

Please check if we do have any side effects with this change.

Comment (halter73) [reply src/Mvc/Mvc.Core/src/ModelBinding/Binders/BodyModelBinder.cs:188]: If we do need to detect the presence of a request body we should use the feature see: https://github.com/dotnet/aspnetcore/pull/39917/files#diff-f137ebb8705f77c3c09cf21ca959b40735794ec71b8fdec0c7d4f0a0b3a7fb34R106

Comment (brunolins16) [reply src/Mvc/Mvc.Core/src/ModelBinding/Binders/BodyModelBinder.cs:188]: Duplicated comment.

Comment (senbar) [reply src/Mvc/Mvc.Core/src/ModelBinding/Binders/BodyModelBinder.cs:188]: Thanks for feedback! I agree with suggestions

**Comment by senbar** [comment]
So I think there should be some sort of ModelState enum in InputFormatterResult class indicating whether model is set null or undefined.

**Comment by senbar** [comment]
I made commit with changes that I think would improve API here. I have added enum describing state of model binding (set/null/empty body). All the logic checking if model is in these states is broken so don't look into that. I haven't made any api changes proposal yet so I dont entirely know how the process goes. That's why I would appreciate here feedback  if this is the way I should finish this issue.

**Comment by brunolins16** [comment]
> I made commit with changes that I think would improve API here. I have added enum describing state of model binding (set/null/empty body). All the logic checking if model is in these states is broken so don't look into that. I haven't made any api changes proposal yet so I dont entirely know how the process goes. That's why I would appreciate here feedback if this is the way I should finish this issue.

@senbar That is an interesting change, however, as you mentioned that will be a large change to how the states are checked. Since the goal is to just provide an more descriptive error message, for JSON "null", but keep rejecting the request as before. I still feel you could try something easier.

If you take a look where the `ReadRequestBodyAsync` is called ([here](https://github.com/dotnet/aspnetcore/blob/1c5da276ce957e9db063d7dc7440ea6cc496300c/src/Mvc/Mvc.Core/src/Formatters/InputFormatter.cs#L95)), it is only called when the body is detected that means the `SystemTextJsonInputFormatter `or `NewtonsoftJsonInputFormatter `will only do the serialization when a body is detected. With that, both formatters will do the serialization and check if the `model is null` and report back the `NoValue` state, that is where I feel we can do better, something like

``` c#
if (model == null && !context.TreatEmptyInputAsDefaultValue)
{
    if ( [check for JSON-encoded value "null"]) 
    { 
        bindingContext.ModelState.AddModelError(context.ModelName, "The new message");
        return InputFormatterResult.Failure();
    } 

    // Some nonempty inputs might deserialize as null, for example whitespace. 
    // The upstream BodyModelBinder needs to
    // be notified that we don't regard this as a real input so it can register
    // a model binding error.
    return InputFormatterResult.NoValue();
}
```

**Comment by senbar** [comment]
To check if request is null here would need reading Request.Body stream in 
```cs
if (model == null && !context.TreatEmptyInputAsDefaultValue)
{
        var bodyStreamReader = new StreamReader(httpContext.Request.Body);
        bodyStreamReader.BaseStream.Seek(0, SeekOrigin.Begin);
        string rawJson = await bodyStreamReader.ReadToEndAsync();
        if(rawJson=="null")
        {
            context.ModelState.TryAddModelError(context.ModelName, "null value not supported");
            return InputFormatterResult.Failure();
        }

        // Some nonempty inputs might deserialize as null, for example whitespace,
        // or the JSON-encoded value "null". The upstream BodyModelBinder needs to
        // be notified that we don't regard this as a real input so it can register
        // a model binding error.
        return InputFormatterResult.NoValue();
}
````
which doesn't work because Request.Body stream is not seekable in general. Any ideas what should I do instead? Buffering whole body stream would have big performance impact.

**Comment by halter73** [comment]
> With that, both formatters will do the serialization and check if the `model is null` and report back the `NoValue` state, that is where I feel we can do better, something like
> 
> ```cs
> if (model == null && !context.TreatEmptyInputAsDefaultValue)
> {
>     if ( [check for JSON-encoded value "null"]) 
>     { 
>         bindingContext.ModelState.AddModelError(context.ModelName, "The new message");
>         return InputFormatterResult.Failure();
>     } 
> 
>     // Some nonempty inputs might deserialize as null, for example whitespace. 
>     // The upstream BodyModelBinder needs to
>     // be notified that we don't regard this as a real input so it can register
>     // a model binding error.
>     return InputFormatterResult.NoValue();
> }
> ```

I'm not sure we want to report failure at such a low layer. After all, `null` is one of the "nonempty inputs might deserialize as null" that the `BodyModelBinder` needs to know about. I don't think that either `!context.TreatEmptyInputAsDefaultValue` nor `EmptyBodyBehavior.Disallow` implies that a `null` literal should be rejected outright. It should be if the parameter is non-nullable, but we don't know that at the `InputFormatter` layer do we?

We probably should be changing this to return `InputFormatterResult.Success(null);` for `null` literals instead of `InputFormatterResult.NoValue();` regardless of `TreatEmptyInputAsDefaultValue` because it's not truly empty.

It's a literal default value, and the whole problem here seems to be that both empty input and null literals are being collapsed into a single `NoValue()` when `TreatEmptyInputAsDefaultValue` is `false`. It's fine for these values to be collapsed when `TreatEmptyInputAsDefaultValue` is `true`, because that's what you asked for, but not the other way around.

I don't think it's necessary to check if there was an empty request body or `null` because the method you're modifying, ` ReadRequestBodyAsync`, is only called when the body is non-empty (or at least we think it will be) that what the `canHaveBody` logic I linked to earlier was doing. It was deciding whether to call `ReadRequestBodyAsync` or return `NoValue()` immediately:

https://github.com/dotnet/aspnetcore/blob/39219394f66854db5d556ad78f9bb065ee48b585/src/Mvc/Mvc.Core/src/Formatters/InputFormatter.cs#L102-L116

And if the `canHaveBody` logic was wrong, and we thought there was going to be a body but it turned out to be empty, `JsonSerializer.DeserializeAsync()` in  `SystemTextJsonInputFormatter` will throw a `JsonException` which will always be turned into a `Failure()`.

Unfortunately Json.NET, turns an empty body into `null` when you try to deserialize. Again the `canHaveBody` checks should prevent this from happening for most empty bodies, but not all. I assume we copied the `if (model == null && !context.TreatEmptyInputAsDefaultValue)` logic from the `NewtonsoftJsonInputFormatter` to the `SystemTextJsonInputFormatter` because it makes very little sense in the latter case where it will only return `null` for a null literal and nothing else. JSON.NET

@brunolins16 Is there anything wrong with just removing the `if (model == null && !context.TreatEmptyInputAsDefaultValue)` branch entirely `SystemTextJsonInputFormatter` and always `return InputFormatterResult.Success(model);` when `JsonSerializer.DeserializeAsync()` doesn't throw?

I imagine we'll have to be more careful with the Json.NET input formatter for back compat. It's always reliably treated `null` and empty bodies as exactly the same because they were indistinguishable from just the call to `Deserialize`. We could in theory try to test if any bytes were actually read from the body stream after the fact to determine the difference, but that would be a breaking change.

**Comment by senbar** [comment]
So isn't this a problem with entire handling of model binding of `"null"` json on nullable parameter with `TreatEmptyInputAsDefaultValue==true`? If we get `"null"` json it gets deserialized to `null` model which returns 
`NoValue` from formatter. This in turn leads to model binding error since:
```cs
// If the input formatter gives a "no value" result, that's always a model state error,
```
From what I understand `TreatEmptyInputAsDefaultValue` refers to empty request body not json `"null"` literal right?

> It should be if the parameter is non-nullable, but we don't know that at the InputFormatter layer do we?

But we do know the type of deserialized object since we are getting it from `context.ModelType` for deserialization in InputFormatter right? Also the same comment I quoted earlier states:
```cs
// If the input formatter gives a "no value" result, that's always a model state error,
// because BodyModelBinder implicitly regards input as being required for model binding.
// If instead the input formatter wants to treat the input as optional, it must do so by
// returning InputFormatterResult.Success(defaultForModelType), because input formatters
// are responsible for choosing a default value for the model type.
```
So I think it is implied that formatter should be handling dealing with optional arguments and stuff like that.

As for checking if json is null literal I was thinking about the case described in the comment: 
`// Some nonempty inputs might deserialize as null, for example whitespace, (...)` 
so I wanted to check if the input we got is actually a null literal or some whitespace. From what I can see know whitespace actually makes deserialization throw `"The input does not contain any JSON tokens. Expected the input to start with a valid JSON token, when isFinalBlock is true. Path: $ | LineNumber: 0 | BytePositionInLine: 1."` . Since the comment made it clear there are _some_ inputs being regarded as null I wanted to make sure the json is null literal (which I am not entirely  sure how to do since request stream is of course non seekable).

**Comment by brunolins16** [comment]
> So isn't this a problem with entire handling of model binding of `"null"` json on nullable parameter with `TreatEmptyInputAsDefaultValue==true`?

I do not think that the condition is wrong just because we did some improvements (.NET 7) in how we are treating the AllowEmptyBody based on the Nullabity in I feel only very specific scenarios will allow `empty body` and not `null` or vice versa.

> Since the comment made it clear there are some inputs being regarded as null I wanted to make sure the json is null literal (which I am not entirely sure how to do since request stream is of course non seekable).

I think the comment about `whitespace` is not true, as you mentioned, and the only situation where we will get a `null` value from the System.Text.Json will be when we have a literal `null`, so i feel we can safe say (only for System.Text.Json) this:

``` csharp
if (model == null && !context.TreatEmptyInputAsDefaultValue)
{
            context.ModelState.TryAddModelError(context.ModelName, "null value not supported");
            return InputFormatterResult.Failure();
}
```

That will produce a better error message, without change the current behavior. And the new mechanism or the user option to allow empty body will be used to decide if the failure will happen or we will treat as `Success` binding.

Also, I think we should keep the `NewtonsoftJsonInputFormatter` as-is for now.

@halter73 anything to add based on our discussion?
@senbar is that make sense for you?

**Comment by senbar** [comment]
I don't see how it may be working correctly for nullable types. When I run 
```cs
 var response = await Client.PostAsJsonAsync<Contact>("/contact", default);
```
model binding returns error even though `Contact` is nullable and null literal was provided. 
The reason is I as I mentioned above: model gets set to null and SystemTextJsonInputFormatter returns NoValue which binder interprets as an error. Isn't this incorrect or do I misunderstand behaviour we are going for?

**Comment by brunolins16** [comment]
> I am not sure if I understand correctly how it deals with nullable types. When I run test with (/contact is simple endpoint with Contact class object argument from body):
> 
> ```cs
>  var response = await Client.PostAsJsonAsync<Contact>("/contact", default);
> ```
> 
> model binding returns error even though `Contact` is nullable and null literal was provided. The reason is I as I mentioned above: model gets set to null and SystemTextJsonInputFormatter returns NoValue which binder interprets as an error. Isn't this incorrect or do I misunderstand behaviour we are going for?

We changed the behavior in .NET 7 and a nullable `Contact` will set the `TreatEmptyInputAsDefaultValue` as `true`. So, a null literal and empty body will be accepted.

**Comment by senbar** [comment]
@brunolins16 I have made suggested changes to SytemTextJsonInputFormatter and added tests for it. 
The question is should I do something with ApiBehaviorTest since it currently relies on SystemTextJsonInputFormatter being used (my null prarameter test would fail if Json.net was used). This may not be the best practise in functional test.

**Comment by brunolins16** [comment]
> @brunolins16 I have made suggested changes to SytemTextJsonInputFormatter and added tests for it. The question is should I do something with ApiBehaviorTest since it currently relies on SystemTextJsonInputFormatter being used (my null prarameter test would fail if Json.net was used). This may not be the best practise in functional test

Let me do some investigation and will reply to you soon. Thanks for your patience and contribution.

**Comment by brunolins16** [comment]
@senbar I just took a look and i think the easiest way is to add the new Functional Tests to https://github.com/brunolins16/aspnetcore/blob/main/src/Mvc/test/Mvc.FunctionalTests/SystemTextJsonInputFormatterTest.cs , instead `APIBehaviorTests`, similar to what we have here:

https://github.com/brunolins16/aspnetcore/blob/e9b9fc0c533b120ce5c0317ee8e3a56aca26d37d/src/Mvc/test/Mvc.FunctionalTests/NewtonsoftJsonInputFormatterTest.cs#L35

File: src/Mvc/test/Mvc.FunctionalTests/ApiBehaviorTest.cs [hunk 2]
```diff
@@ -82,6 +83,73 @@ public virtual async Task ActionsReturnBadRequest_WhenModelStateIsInvalid()
...
+                {
+                    var error = Assert.Single(kvp.Value);
+                    Assert.Equal("null value not supported", error);
+                }
+            );
+        }
+    }
+
+    [Fact]
+    public virtual async Task ActionWithNonOptionalParameterReturnsBodyCannotBeEmpty_WhenInvokedWithEmptyBody()
```
Comment (brunolins16) [inline src/Mvc/test/Mvc.FunctionalTests/ApiBehaviorTest.cs:120]:
```suggestion
    public Task JsonInputFormatter_ReturnsBadRequest_ForEmptyRequestBodyAction()
```

Comment (brunolins16) [reply src/Mvc/test/Mvc.FunctionalTests/ApiBehaviorTest.cs:120]: As mentioned, maybe you should move it to https://github.com/brunolins16/aspnetcore/blob/main/src/Mvc/test/Mvc.FunctionalTests/SystemTextJsonInputFormatterTest.cs

File: src/Mvc/test/Mvc.FunctionalTests/ApiBehaviorTest.cs [hunk 3]
```diff
@@ -82,6 +83,73 @@ public virtual async Task ActionsReturnBadRequest_WhenModelStateIsInvalid()
         }
     }
 
+    [Fact]
+    public virtual async Task ActionWithNonOptionalParameterReturnsNullIsNotPermited_WhenInvokedWithNullParameter()
```
Comment (brunolins16) [inline src/Mvc/test/Mvc.FunctionalTests/ApiBehaviorTest.cs:87]:
```suggestion
    public Task JsonInputFormatter_ReturnsBadRequest_ForLiteralNull()
```

**Comment by senbar** [comment]
Sorry have been busy last weeks, I will be finishing PR soon

**Comment by brunolins16** [comment]
> Sorry have been busy last weeks, I will be finishing PR soon

@senbar are you still planning to work on it? Can I help you on anything?

**Comment by senbar** [comment]
Yeah sorry aobut that I was changing job lately and was swamped with work. I will try to finish it next two weeks. As for help I think I got it it should be pretty straightforward, if there are any problems I will let You know, thanks

**Comment by brunolins16** [comment]
Don't worry about it,  take your time. I was just checking if I could do anything to help.

**Pull request closed without merging**